### PR TITLE
Generic linear solving

### PR DESCRIPTION
**Pull request by joschmitt**
I started working on a unified interface for linear solving (in context with OSCAR 1.0).
@thofma is this the interface you had in mind?

(I will add tests and the functionality for rings.)

### CODE REVIEW

**Comment by thofma** [comment]
Yes, this looks about right. I wonder whether we should have some (internal) lazy transpose wrapper to ease the pain with the left/right business. Let's discuss it on Friday.

File: src/Solve.jl [hunk 1]
```diff
@@ -81,65 +125,67 @@ function _can_solve_internal(A::MatElem{T}, b::MatElem{T}, task::Symbol; side::S
...
-  else
-    ncols(A) != ncols(b) && error("Incompatible matrices")
-    mu = _hcat_transposed(A, b)
-    # For side == :left we pretend that A and b are transposed!
-    ncolsA = nrows(A)
-    ncolsb = nrows(b)
+  if isleft
+    # For side == :left, we pretend that A and b are transposed
+    A = lazy_transpose(A)
+    b = lazy_transpose(b)
```
Comment (lgoettgens) [inline src/Solve.jl:135]: overwriting these already existing variables with something of a different type is unfortunately type unstable

**Comment by joschmitt** [comment]
I now added all the functionality we wanted (I think). Assuming the tests pass, this is good to go from my perspective.

**Comment by lgoettgens** [comment]
At the moment, this only adds new functionality in a submodule, right?
I wonder if replacing the current mix of functions with this new stuff will break anything downstream.

**Comment by joschmitt** [comment]
> At the moment, this only adds new functionality in a submodule, right? I wonder if replacing the current mix of functions with this new stuff will break anything downstream.

Yes, that's going to be the fun part. I think @thofma had a plan how to make these changes available without breaking anything (that's why we have the submodule for the moment).

**Comment by joschmitt** [comment]
Sounds good to me. Let's see how it goes.

**Review by thofma — APPROVED** [review]

**Comment by fieker** [comment]
On Thu, Jan 18, 2024 at 07:40:35AM -0800, Tommy Hofmann wrote:
> Here was my plan.
> 1. Merge.
> 2. Do a quick minor release
> 3. Implement the interface for the various matrix types in Nemo
> 4. Then we have two options. Do (inside Hecke) `solve = Solving.solve` etc. or replace all `solve` by `Solving.solve` and check how much breaks.
> 5. Repeat step 4 with Oscar.
> 6. Go back to AA and replace the old `AA.solve` stuff.
> 
> Does that make sense?
> 
Sounds like work, but possibly the best bet we have

**Pull request merged**